Post: Performance Report Summer 2013

xiongchiamiov opened the pull request:
Etsy and others have recently been releasing very candid reports of their
sites' performance. Inspired by this, and wanting to both spur ourselves onward
and to give others a view of how similar sites are doing, we present to you
iFixit's first performance report.

xiongchiamiov committed:
Post: Performance Report Summer 2013

Etsy and others have recently been releasing very candid reports of their
sites' performance. Inspired by this, and wanting to both spur ourselves onward
and to give others a view of how similar sites are doing, we present to you
iFixit's first performance report.

danielbeardsley commented:
I'd vote for excluding the Admin stuff, it's not really relevant to end-users and i could see folks mistaking 'Admin' for 'Site Management'

Also, you should double-check the math on the response-time graphs, I'm not seeing the same thing at all for `response_time.mean` on that day. Unless you're graphing `mean_90` or something.

danielbeardsley commented on _posts/2013-06-16-performance-report.md:
```diff
@@ -0,0 +1,132 @@
+---
+layout: with-comments
+title: "Performance Report: Summer 2013"
```
Perhaps `Website Performance Report: ...` or `Website Speed report: ...`  I dunno, I feel like performance is ambigous... but ... this is a Dev blog, so maybe it's obvious.

danielbeardsley commented on _posts/2013-06-16-performance-report.md:
```diff
@@ -0,0 +1,132 @@
+| Answers | 233.92 | 243.19 | 102.22         | 79.15 |  6466.81
+| Search  | 134.64 | 184.20 | 257.93         | 38.02 | 10752.06
+
+This statistics are recorded inside our application, so let's grab a view from
```
These

danielbeardsley commented on _posts/2013-06-16-performance-report.md:
```diff
@@ -0,0 +1,132 @@
+where it is rolled up into 10-second groupings.  Let's take a look at some
+graphs of response times from last Friday, 2013-06-14.
+
+![overall](/assets/2013-06-16-performance-report/overall.png)
```
It might be useful to show a longer term graph, or... a table instead, graphs are pretty, but the values aren't really changing over time much.

xiongchiamiov replied:
Yeah, I wasn't sure about that.  The graph's kinda useful to show that it _isn't_ changing much, but probably we can revamp the table below to include the breakdown and/or do some better charting.  I'm open to suggestions on what's the most useful.

danielbeardsley commented on _posts/2013-06-16-performance-report.md:
```diff
@@ -0,0 +1,132 @@
+Here are the averages calculated from Graphite across the whole day, using
+Graphite's CSV output, a little unix shell magic, and [the average project]:
+
+| Section | Median | Mean   | Std Deviation  | Min   | Max
```
How are you calculating these aggregates? The mean is pretty easy using `response_time.sum` and `.count`, but I wouldn't know how to go about aggregating the 10-sec interval standard deviation to something that represents all day, not to mention median.

Also, is this the `min` and `max` of the 10s means? or of over all?

I don't think min and max are nearly as useful as a 90% interval (`upper_90`)

xiongchiamiov replied:
```
[$]> cat gather
#!/bin/bash -e

http "http://graphite.endpoint.com/render?from=00%3A00_20130614&until=23%3A30_20130614&target=$1&format=csv" |
awk -F , '{print $3}' |
average
```

The smallest granularity we store in Graphite is 10s means, which is usually good enough, but could be a little misleading in this situation.  I originally tried doing the median through graphite, but it's hard to get the endpoints and the range to line up, so it was easier to just pull all the data out through the text api and calculate it elsewhere.

We should be able to do the 90% interval without too much trouble.

danielbeardsley replied:
Yeah, that script makes sense, but what does the 'average' std deviation of the 10s intervals really mean? i.e. This is the mean of the std-deviations and the mean of the maxes and mins, it's kinda weird. Mean of means is probably really close to the true mean, and the mean of std-devs is probably close enough, but min / max seem odd in that it's the mean of the minimums of each 10s interval.

Anyway, stats are never perfect, but I'd vote for dropping the min / max and use the mean of `mean_90`

marczych commented on _posts/2013-06-16-performance-report.md:
```diff
@@ -0,0 +1,132 @@
+Thankfully, search requests somewhat make up for Answers, running almost
+30-50ms faster than the average.  This is thanks in part to having a much
+higher percentage of AJAX calls versus full page renders, as well as the
+replacement of Sphinx for MySQL as the primary datasource.
```
[data source?](http://en.wikipedia.org/wiki/Data_source)

adam000 replied:
^agreed.

marczych commented on _posts/2013-06-16-performance-report.md:
```diff
@@ -0,0 +1,132 @@
+| Generate Sizes |   2.92 | 2.23 | 1.21           | 0.25 |   5.89
+
+While the new system [Cedric] just put into place improved our markup rendering
+by an order of magnitude, it still seems high, both in the the max of almost
```
the the

marczych commented:
Looks like there are a few images that aren't used (bot.png, cacheable.png, uncacheable.png). Did you commit them just so we have the graphs even though they aren't used in the write-up.

adam000 commented:
How might I view this in-browser? I'm curious to see the output but running it through Markdown got strange results:

![](http://f.cl.ly/items/0T1i321J2b3h3a1r3n2H/Screen%20Shot%202013-06-18%20at%2011.47.20%20AM.png)

(yes, the images are an absolute path so that could be fixed but the title stuff... ick)

adam000 commented:
Cool stuff! Just reading it from command line, looks good.

Text Reviewed, :+1:

danielbeardsley commented:
> How might I view this in-browser?

This uses the jekyll site generator.

```
$ gem install jekyll
$ git clone https://github.com/iFixit/itbrokeand.ifixit.com
$ cd itbrokeand.ifixit.com
$ jeykll #generates the site and starts a server on port 4000
```

xiongchiamiov commented:
> I'd vote for excluding the Admin stuff, it's not really relevant to end-users and i could see folks mistaking 'Admin' for 'Site Management'

Yeah, you're right.  It's useful to us, internally, which is really part of the audience, but I guess I should just keep that stuff for a separate report.

> Also, you should double-check the math on the response-time graphs, I'm not seeing the same thing at all for response_time.mean on that day. Unless you're graphing mean_90 or something.

Take a look at the 'Performance Report' dashboard.

> Looks like there are a few images that aren't used (bot.png, cacheable.png, uncacheable.png). Did you commit them just so we have the graphs even though they aren't used in the write-up.

I made a number of graphs and only ended up using some of them, and I was silly and just added the whole folder.  Removed.

xiongchiamiov committed:
Fix typos.

xiongchiamiov committed:
Don't talk about Admin.

It's internal, and I guess we're trying to focus on more user-facing metrics.

xiongchiamiov commented:
I fixed the typos and removed the stuff about Admin.

xiongchiamiov commented:
I realized I used `mean_90` on those graphs (but not the tabled averages).  Silly me.

marczych commented:
@xiongchiamiov Do we still want to publish this?

xiongchiamiov commented:
Nah, I (or we) should rewrite this, using new data, and knowing new stuff about performance, and at that point we might as well start (pretty much) from scratch.
